src/Juvix/Core/FromFrontend.hs

```diff
@@ -0,0 +1,260 @@
+  | NotAGUsage FE.Expression
+
+data CoreSig' ext primTy primVal =
+    DataSig (IR.Term' ext primTy primVal)
+  | ValSig  IR.GlobalUsage (IR.Term' ext primTy primVal)
+  deriving Generic
+
+deriving instance
+  ( Eq primTy, Eq primVal,
+    IR.TermAll Eq ext primTy primVal,
+    IR.ElimAll Eq ext primTy primVal
+  ) =>
+  Eq (CoreSig' ext primTy primVal)
+
+deriving instance
+  ( Show primTy, Show primVal,
+    IR.TermAll Show ext primTy primVal,
+    IR.ElimAll Show ext primTy primVal
+  ) =>
+  Show (CoreSig' ext primTy primVal)
+
+deriving instance
+  ( Data ext, Data primTy, Data primVal,
+    IR.TermAll Data ext primTy primVal,
+    IR.ElimAll Data ext primTy primVal
+  ) =>
+  Data (CoreSig' ext primTy primVal)
+
+type CoreSigIR = CoreSig' IR.NoExt
+type CoreSigHR = CoreSig' HR.T
+
+type CoreSigs' ext primTy primVal =
+  HashMap IR.GlobalName (CoreSig' ext primTy primVal)
+
+type CoreSigsIR primTy primVal = CoreSigs' IR.NoExt primTy primVal
+type CoreSigsHR primTy primVal = CoreSigs' HR.T     primTy primVal
+
+
+data FFState primTy primVal =
+  FFState {
+    frontend :: FE.FinalContext,
+    param    :: P.Parameterisation primTy primVal,
+    coreSigs :: CoreSigsIR primTy primVal,
+    core     :: IR.Globals primTy primVal
+  }
+  deriving Generic
+
+type EnvAlias primTy primVal =
+  ExceptT Error (State (FFState primTy primVal))
+
+newtype Env primTy primVal a =
+    Env {unEnv :: EnvAlias primTy primVal a}
+  deriving newtype (Functor, Applicative, Monad)
+  deriving (HasThrow "fromFrontendError" Error)
+    via MonadError (EnvAlias primTy primVal)
+  deriving
+    ( HasSource "frontend" FE.FinalContext,
+      HasReader "frontend" FE.FinalContext
+    )
+    via ReaderField "frontend" (EnvAlias primTy primVal)
+  deriving
+    ( HasSource "param" (P.Parameterisation primTy primVal),
+      HasReader "param" (P.Parameterisation primTy primVal)
+    )
+    via ReaderField "param" (EnvAlias primTy primVal)
+  deriving
+    ( HasSource "core" (IR.Globals primTy primVal),
+      HasSink   "core" (IR.Globals primTy primVal),
+      HasState  "core" (IR.Globals primTy primVal)
+    )
+    via StateField "core" (EnvAlias primTy primVal)
+
+throwFF :: Error -> Env primTy primVal a
+throwFF = throw @"fromFrontendError"
+
+
+paramConstant :: P.Parameterisation primTy primVal
+              -> FE.Constant -> Maybe primVal
+paramConstant p (FE.Number (FE.Double'  d)) = P.floatVal  p d
+paramConstant p (FE.Number (FE.Integer' i)) = P.intVal    p i
+paramConstant p (FE.String (FE.Sho      s)) = P.stringVal p s
+
+
+transformTerm :: FE.Expression -> Env primTy primVal (IR.Term primTy primVal)
+transformTerm = fmap hrToIR . transformTermHR
+
+transformTermHR :: FE.Expression -> Env primTy primVal (HR.Term primTy primVal)
+transformTermHR (FE.Constant k) = do
+  p <- ask @"param"
+  case paramConstant p k of
+    Just x  -> pure $ HR.Prim x
+    Nothing -> throwFF $ UnsupportedConstant k
+transformTermHR (FE.Let l) = _
```

Comment:
Presumably we can turn this into core `let`?

Reply:
Frontend lets can be pattern matching functions, so their body will need to be lifted I think.

But if it's just `let foo a b c = ... in` then it can be transformed to `let foo = λ a b c → ... in` as a special case

Reply:
I presume once we have case, we can forgo lambda lifting and write something akin to `let foo =  λ a b c → case (a,b,c,) of ...`

Reply:
Yeah, exactly. It's just that would unfortunately be a very invasive change to core